src/main/drivers/vtx_table.c

```diff
@@ -0,0 +1,121 @@
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Created by jflyper */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <ctype.h>
+#include <string.h>
+
+#include "platform.h"
+
+#if defined(USE_VTX_TABLE)
+
+#include "common/printf.h"
+
+#include "pg/pg.h"
+#include "pg/vtx_table.h"
+#include "drivers/vtx_common.h"
+#include "drivers/vtx_table.h"
+
+int            vtxTableBandCount;
+int            vtxTableChannelCount;
+uint16_t       vtxTableFrequency[VTX_TABLE_MAX_BANDS][VTX_TABLE_MAX_CHANNELS];
+const char *   vtxTableBandNames[VTX_TABLE_MAX_BANDS + 1];
+char           vtxTableBandLetters[VTX_TABLE_MAX_BANDS + 1];
+const char *   vtxTableChannelNames[VTX_TABLE_MAX_CHANNELS + 1];
+
+int            vtxTablePowerLevels;
+uint16_t       vtxTablePowerValues[VTX_TABLE_MAX_POWER_LEVELS];
+const char *   vtxTablePowerLabels[VTX_TABLE_MAX_POWER_LEVELS + 1];
+
+void vtxTableInit(void)
+{
+    const vtxTableConfig_t *config = vtxTableConfig();
+
+    for (int band = 0; band < VTX_TABLE_MAX_BANDS; band++) {
+        for (int channel = 0; channel < VTX_TABLE_MAX_CHANNELS; channel++) {
+            vtxTableFrequency[band][channel] = config->frequency[band][channel];
+        }
+        vtxTableBandNames[band + 1] = config->bandNames[band];
+        vtxTableBandLetters[band + 1] = config->bandLetters[band];
+    }
+
+    vtxTableBandNames[0] = "--------";
+    vtxTableBandLetters[0] = '-';
+
+    for (int channel = 0; channel < VTX_TABLE_MAX_CHANNELS; channel++) {
+        vtxTableChannelNames[channel + 1] = config->channelNames[channel];
+    }
+    vtxTableChannelNames[0] = "-";
+
+    for (int level = 0; level < config->powerLevels; level++) {
+        vtxTablePowerValues[level] = config->powerValues[level];
+        vtxTablePowerLabels[level + 1] = config->powerLabels[level];
+    }
+    vtxTablePowerLabels[0] = "---";
+
+    vtxTablePowerLevels = config->powerLevels;
+}
+
+void vtxTableFillVtxDevice(vtxDevice_t *device)
+{
+    device->capability.bandCount = VTX_SETTINGS_BAND_COUNT;
+    device->capability.channelCount = VTX_SETTINGS_CHANNEL_COUNT;
+    device->capability.powerCount = vtxTablePowerLevels;
+    device->frequencyTable = (uint16_t *)vtxTableFrequency;
+    device->bandNames = (char **)&vtxTableBandNames;
+    device->channelNames = (char **)&vtxTableChannelNames;
+    device->powerNames = (char **)&vtxTablePowerLabels;
+}
+
+void vtxTableStrncpyWithPad(char *dst, char *src, int length)
```

Comment:
maybe `sprintf(dst, "%-*.*s", length, length, src);`, but I'm not sure if BF printf does support this

Reply:
Nah, I did found out both left justification and padding are not supported 😭

Reply:
Yes, please! I'm using `vtxTableStrncpyWithPad` everywhere now!